pkg/shell/base_index.js

```diff
@@ -496,7 +498,84 @@ function Index() {
+    let session_final_timer = null;
+    let session_timeout = 0;
+    let final_countdown = 30000; // last 30 seconds
+    let title = "";
+
+    function startTimer() {
+        if (session_timeout > 0)
+            session_timer = window.setTimeout(sessionTimeout, session_timeout - final_countdown);
+    }
+
+    function sessionTimeout() {
+        /* When non-password login is in use, we don't want to logout used, as the user can just log
+         * back in. This can be recognized by checking `privileges` but this is set up
+         * asynchronously and it may happen that it is not yet set up in StartTimer().
+         * We can just check this out on first sessionTimeout() and if non-pwd login is used then
+         * just remove the timer and do not bother with it anymore.
+         */
+        if (self.privileges.is_locked() == true) {
+            window.clearTimeout(session_timer);
+            return;
+        }
+
+        title = document.title;
+        sessionFinalTimeout();
+        $("#session-timeout-dialog").modal("show");
+
+        document.getElementById("keep-session-alive").addEventListener("click", e => {
+            window.clearTimeout(session_final_timer);
+            session_final_timer = null;
+            document.title = title;
+            resetTimer();
+            $("#session-timeout-dialog").modal("hide");
+            final_countdown = 30000;
+        });
+    }
+
+    function updateFinalCountdown() {
+        const remaining_secs = Math.floor(final_countdown / 1000);
+        const timeout_text = cockpit.format(_("You will be logged out in $0 seconds."), remaining_secs);
+        document.getElementById("timeout-message").innerHTML = timeout_text;
+        document.title = "(" + remaining_secs + ")" + title;
+    }
+
+    function sessionFinalTimeout() {
+        final_countdown -= 1000;
+        if (final_countdown > 0) {
+            updateFinalCountdown();
+            session_final_timer = window.setTimeout(sessionFinalTimeout, 1000);
+        } else {
+            cockpit.logout(true, _("You have been logged out due to inactivity."));
+        }
+    }
+
+    function resetTimer() {
+        if (!session_final_timer) {
+            window.clearTimeout(session_timer);
+            startTimer();
+        }
+    }
+
+    function setupTimers(document) {
+        document.addEventListener("mousemove", resetTimer, false);
+        document.addEventListener("mousedown", resetTimer, false);
+        document.addEventListener("keypress", resetTimer, false);
+        document.addEventListener("touchmove", resetTimer, false);
+        document.addEventListener("onscroll", resetTimer, false);
+    }
+
+    setupTimers(window);
+    cockpit.dbus(null, { bus: "internal" }).call("/config", "cockpit.Config", "GetUInt", ["Session", "IdleTimeout", 15, 240, 0], [])
+            .then(result => {
+                session_timeout = result[0] * 60000;
+                startTimer();
+            });
```

Comment:
Let's please .catch() exceptions here. It may be that we are running against an older bridge that doesn't yet have this API. In that case (please check the error code), let's just ignore the error, otherwise console.warn() about it.